### ISSUE TITLE
Type of numpy.array is not properly inferred

### ISSUE THREAD

**Issue by astropenguin** [microsoft/pyright#1717]
**Describe the bug**

With the latest Pyright with the strict mode, `reportUnknownMemberType` is raised when I use `numpy.array()` or array-creation functions (like `numpy.ones()`). It does not occur when I use from-style import like `from numpy import array` or when I use older versions of Pyright (<1.1.124).

**To Reproduce**

Write a code like:

```python
import numpy


arr = numpy.array(...)  # -> Type of "array" is partially unknown
```

with any of the following conditions:

- NumPy version: 1.20 or higher
- Pyright version: 1.1.124 or higher
- Python version: 3.7, 3.8, or 3.9
- Type check mode: "strict"

**Expected behavior**

With both normal-style and from-style import, the error should not be raised.

**Screenshots or Code**

- [Minimum reproducible example (astropenguin/pyright-numpy-bug)](https://github.com/astropenguin/pyright-numpy-bug)
- [Tests by GitHub Actions (astropenguin/pyright-numpy-bug/actions)](https://github.com/astropenguin/pyright-numpy-bug/actions/runs/711684415)

**Comment by astropenguin** [comment]
This might be a kind of bug of NumPy, however, I opened the issue here because it depends on the version of Pyright, not that of NumPy. Please feel free to close it if it is not in a right place. Thank you!

**Comment by erictraut** [comment]
Thanks for reporting the issue — and for the excellent repro steps. Here's what's going on.

A minor terminology correction... there's no "type inference" occurring in this case because all of the types involved are declared. Type "inference" refers to the situation where types are not declared and the type checker needs to evaluate the type of an expression based on available contextual information.

Let me first address the apparent inconsistency in error reporting. The error is because of a diagnostic check called "reportUnknownMemberType". It's applied to member access expressions (of the form `a.b`). It validates that the type of member `b` is fully known. This diagnostic check does not apply to expressions that contain no member access. So that explains why `array` doesn't generate an error but `numpy.array` does. The "reportUnknownMemberType" diagnostic check is off by default in pyright, but it can be enabled explicitly in the configuration. It's also on by default when "strict" mode is enabled.

Next, let's look at why the type of the `array` function is considered "partially unknown". The second parameter of this function has a declared type of `DTypeLike`. This is a type alias that includes the type `dtype`. The `dtype` class is generic and expects to receive a single type argument, but no type argument is provided in this case. I'm guessing that the numpy maintainers mean for it to be `dtype[Any]`. The correct fix is to add that missing type argument in the `numpy/typing/_dtype_like.py` file. I recommend filing a bug or a PR with numpy.

Using string type checking with numpy today is going to be a challenge because there are many holes in numpy's type information. The CLI version of pyright contains a "--verifytypes" option that allows you to determine the "type completeness" of a library. Numpy's type completeness is currently at 14.6%, so you're likely going to hit unknown or partially-unknown types.

Here are some workarounds:
1. Disable strict mode and manually enable only those diagnostic checks that make sense for your project.
2. Use file-level "# pyright: reportUknownMemberType=false" comments to suppress that diagnostic check within your file.
3. For the specific case of "numpy.array", use the `from numpy import array` trick that you used in your sample code. This avoids the reportUnknownMemberType check.
4. Create your own type stubs for numpy for the subset of functionality that you use. Place them in a subdirectory within your project, and use the stubPath setting to point to them. (This is the technique we use on my team.) It's a lot of work to create and maintain your own type stubs, so this option isn't for the faint of heart. You can generate a "first draft" of these stubs using the [pyright CLI](https://github.com/microsoft/pyright/blob/master/docs/command-line.md) with the "--createstub" option.
